dbms/src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -293,20 +310,53 @@ void StorageDistributedDirectoryMonitor::readQueryAndSettings(
                     << "It may lack support for new features.");
         }
 
+        /// Verify checksums
+        if (initiator_revision >= DBMS_MIN_REVISION_WITH_EXTRAS_CHECKSUM_IN_DIST_BATCH)
+        {
+            CityHash_v1_0_2::uint128 expected_checksum;
+            CityHash_v1_0_2::uint128 calculated_checksum;
+            size_t checksum_size = sizeof(expected_checksum)*2;
+
+            if (extra_info.available() < checksum_size)
+                throw Exception("Not enough data", ErrorCodes::CORRUPTED_DATA);
+
+            ReadBuffer checksums(extra_info.position(), extra_info.available(), extra_info.available() - checksum_size);
+            extra_info = ReadBufferFromMemory(extra_info.position(), extra_info.available() - checksum_size);
```

Comment:
Recreating `extra_info` in the local scope looks complicated. It seems to be easier to create two different ReadBufferFromMemory from the beginning like this:

```
        String extra_info_with_checksum;
        readStringBinary(extra_info_with_checksum, in);
        ReadBufferFromMemory extra_info(extra_info_with_checksum.data(), extra_info_with_checksum.size() - checksum_size);
        ReadBufferFromMemory extra_info_checksum(extra_info_with_checksum.data() + extra_info_with_checksum.size() - checksum_size, checksum_size);
```

Reply:
>Recreating extra_info in the local scope looks complicated

Agree

>It seems to be easier to create two different ReadBufferFromMemory from the beginning like this:

But AFAICS it is not possible, since:
- we need to get revision first (to know is there any checksum info)
- `ReadBuffer::set` reset the `working_buffer`, so it not the analog of `seek`/`truncate`, it is more kind of `reset`

But if we do not asserts for the `eof` then we can simply leave it as-is and create another buffer just for checking checksum, in the local scope, although this does not looks cleaner

Reply:
>Recreating extra_info in the local scope looks complicated

And also not only complicated, but does not [allowed](https://clickhouse-builds.s3.yandex.net/7914/a1d8df212c1f3ad1e9c4bf1c80269d203c401938/build_log_557681232_1574981153.txt):

```
2019-11-28 22:15:51 ../dbms/src/IO/ReadBuffer.h:69:13: error: definition of implicit copy assignment operator for 'ReadBuffer' is deprecated because it has a user-declared destructor [-Werror,-Wdeprecated]
2019-11-28 22:15:51     virtual ~ReadBuffer() {}
2019-11-28 22:15:51             ^
2019-11-28 22:15:51 ../dbms/src/IO/ReadBufferFromMemory.h:13:7: note: in implicit copy assignment operator for 'DB::ReadBuffer' first required here
2019-11-28 22:15:51 class ReadBufferFromMemory : public ReadBuffer
2019-11-28 22:15:51       ^
2019-11-28 22:15:51 ../dbms/src/Storages/Distributed/DirectoryMonitor.cpp:324:24: note: in implicit move assignment operator for 'DB::ReadBufferFromMemory' first required here
2019-11-28 22:15:51             extra_info = ReadBufferFromMemory(extra_info.position(), extra_info.available() - checksum_size);
2019-11-28 22:15:51                        ^

```

Reply:
> > Recreating extra_info in the local scope looks complicated
> 
> Agree
> 
> > It seems to be easier to create two different ReadBufferFromMemory from the beginning like this:
> 
> But AFAICS it is not possible, since:
> 
> * we need to get revision first (to know is there any checksum info)
> * `ReadBuffer::set` reset the `working_buffer`, so it not the analog of `seek`/`truncate`, it is more kind of `reset`
> 
> But if we do not asserts for the `eof` then we can simply leave it as-is and create another buffer just for checking checksum, in the local scope, although this does not looks cleaner

The current version is still [54429](https://github.com/ClickHouse/ClickHouse/blob/master/dbms/cmake/version.cmake#L2). `DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO` was introduced in the same version. So you don't need to check that `initiator_revision >= DBMS_MIN_REVISION_WITH_EXTRAS_CHECKSUM_IN_DIST_BATCH`. If `query_size_or_magic` is equal to `DBMS_DISTRIBUTED_SIGNATURE_EXTRA_INFO` you can be sure that the current version supports checksums. This fact can be used to simplify the conditions.
```